cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -397,6 +397,136 @@ struct PowerChecked {
+    return std::asin(val);
+  }
+
+  template <typename T, typename Arg0>
+  static enable_if_floating_point<Arg0, T> Call(KernelContext*, Arg0 val, Status*) {
+    static_assert(std::is_same<T, Arg0>::value, "");
+    return std::asin(val);
+  }
+};
+
+struct TrigAsinChecked {
+  template <typename T, typename Arg0>
+  static enable_if_integer<Arg0, T> Call(KernelContext*, Arg0 val, Status*) {
+    static_assert(std::is_same<T, double>::value, "");
+    return std::asin(val);
+  }
+
+  template <typename T, typename Arg0>
+  static enable_if_floating_point<Arg0, T> Call(KernelContext*, Arg0 val, Status* st) {
+    static_assert(std::is_same<T, Arg0>::value, "");
+    if (ARROW_PREDICT_FALSE((val < -1.0 || val > 1.0) && !std::isnan(val))) {
+      *st = Status::Invalid("domain error");
+      return val;
+    }
+    return std::asin(val);
+  }
+};
+
+struct TrigAcos {
+  template <typename T, typename Arg0>
+  static enable_if_integer<Arg0, T> Call(KernelContext*, Arg0 val, Status*) {
+    static_assert(std::is_same<T, double>::value, "");
+    return std::acos(val);
+  }
+
+  template <typename T, typename Arg0>
+  static enable_if_floating_point<Arg0, T> Call(KernelContext*, Arg0 val, Status*) {
+    static_assert(std::is_same<T, Arg0>::value, "");
+    return std::acos(val);
+  }
+};
+
+struct TrigAcosChecked {
+  template <typename T, typename Arg0>
+  static enable_if_integer<Arg0, T> Call(KernelContext*, Arg0 val, Status*) {
+    static_assert(std::is_same<T, double>::value, "");
+    return std::acos(val);
+  }
+
+  template <typename T, typename Arg0>
+  static enable_if_floating_point<Arg0, T> Call(KernelContext*, Arg0 val, Status* st) {
+    static_assert(std::is_same<T, Arg0>::value, "");
+    if (ARROW_PREDICT_FALSE((val < -1.0 || val > 1.0) && !std::isnan(val))) {
+      *st = Status::Invalid("domain error");
+      return val;
+    }
+    return std::acos(val);
+  }
+};
+
+struct TrigAtan {
+  template <typename T, typename Arg0>
+  static enable_if_integer<Arg0, T> Call(KernelContext*, Arg0 val, Status*) {
+    static_assert(std::is_same<T, double>::value, "");
+    return std::atan(val);
+  }
+
+  template <typename T, typename Arg0>
+  static enable_if_floating_point<Arg0, T> Call(KernelContext*, Arg0 val, Status*) {
+    static_assert(std::is_same<T, Arg0>::value, "");
+    return std::atan(val);
+  }
+};
+
+struct TrigAtan2 {
+  template <typename T, typename Arg0, typename Arg1>
+  static enable_if_integer<Arg0, T> Call(KernelContext*, Arg0 y, Arg1 x, Status*) {
+    static_assert(std::is_same<T, double>::value, "");
+    static_assert(std::is_same<Arg0, Arg1>::value, "");
+    return std::atan2(y, x);
+  }
+
+  template <typename T, typename Arg0, typename Arg1>
+  static enable_if_floating_point<Arg0, T> Call(KernelContext*, Arg0 y, Arg1 x, Status*) {
+    static_assert(std::is_same<T, Arg0>::value, "");
+    static_assert(std::is_same<Arg0, Arg1>::value, "");
+    return std::atan2(y, x);
+  }
+};
```

Comment:
Are we missing the `TrigAtanChecked` and `TrigAtan2Checked`?

Reply:
I was going by cppreference which states that these functions don't have domain errors. The SEI CERT C coding standard disagrees in the case of atan2 (it states x and y must not be 0), but after digging up IEEE754, it states that atan2(0, 0) is defined (as 0).

These functions do have range (overflow) errors though so I can add checked versions for that.

Reply:
Actually, I take that back: looking at IEEE754 neither of them are specified to overflow. They can both underflow, but this isn't undefined behavior/I don't think we need to check for that?

Reply:
But do we want to add the ArithmeticOptions to the signature anyways for consistency?

Reply:
C++ defines atan2(0, 0) as a domain error, the returned value is implementation-specific, so even if it returns 0 in our tests we should trigger a domain error as it maps to 0/0 expression.

Reply:
What I see is that an implementation using IEEE floating point arithmetic specifically does not consider it a domain error: https://en.cppreference.com/w/cpp/numeric/math/atan2

Though we could be more conservative.

Reply:
I have always thought of the `check_overflow` option to be applicable for both *underflow* and *overflow*. I do not see a benefit of having two distinct options. Regarding if you should include `ArithmeticOptions` just for the sakes of consistency, I would argue that should not be done. Why would we want to include an unused value? Also, it might confuse the client code if it thinks that the function options provide some functionality.

Reply:
Thanks for the `atan2` link. Maybe a safer way to go (to prevent domain error for `atan2(0, 0)`) is to explicitly return 0 for this case. This way if IEEE 754 is not being conformed to, result is consistent and domain error is circumvented.

Reply:
Regarding underflow: I will adjust the kernels to check for this and see if I can find some values to test with.

Reply:
Well, probably there is no need to check for underflow, because it seems that for all trigonometric functions, C++ applies rounding to the underflow cases and returns a correct value.

Reply:
Ah, ok. I do see that it'll set a floating-point exception that you can test for, but that may indeed be overkill.

Reply:
I've fixed the docs and updated atan2 to explicitly check for the 0, 0 case (including signed zeroes which are specified).

Reply:
One last fix, `copy_sign` should also be applied to 0.0 value, `return std::copysign(static_cast<T>(0.0), y)`.

Reply:
Whoops, good catch - thanks!